### PR TITLE
Add user shops

### PR DESCRIPTION
**Pull request by 0Lucifer0**

### CODE REVIEW

**Comment by 0Lucifer0** [comment]
Will add the tests in another pr

**Comment by 0Lucifer0** [comment]
fix #408

File: src/NosCore.Controllers/NpcPacketController.cs [hunk 1]
```diff
@@ -135,10 +155,144 @@ public void Shopping(ShoppingPacket shoppingPacket)
...
+        public void CreateShop(MShopPacket mShopPacket)
+        {
+            if (Session.Character.InExchangeOrTrade)
+            {
+                //todo log
+                return;
+            }
+
+            if (Session.Character.MapInstance.Portals.Any(por
+                => Session.Character.PositionX < por.SourceX + 6
```
Comment (Elendan) [inline src/NosCore.Controllers/NpcPacketController.cs]: Making a method to check in a circle could be useful for later (quests for example)

File: src/NosCore.Controllers/NpcPacketController.cs [hunk 2]
```diff
@@ -135,10 +155,144 @@ public void Shopping(ShoppingPacket shoppingPacket)
...
+                    {
+                        Session.SendPacket(Session.Character.GenerateSay(
+                            Language.Instance.GetMessageFromKey(LanguageKey.SHOP_EMPTY, Session.Account.Language),
+                            SayColorType.Yellow));
+                        Session.SendPacket(new ShopEndPacket { Type = 0 });
+                        return;
+                    }
+
+                    Session.Character.Shop.Session = Session;
+                    Session.Character.Shop.MenuType = 3;
```
Comment (Elendan) [inline src/NosCore.Controllers/NpcPacketController.cs]: Maybe an enum for menutypes ?

File: src/NosCore.GameObject/ComponentEntities/Extensions/AliveEntityExtension.cs [hunk 3]
```diff
@@ -258,6 +258,18 @@ public static RestPacket GenerateRest(this IAliveEntity aliveEntity)
...
+        public static PflagPacket GeneratePFlag(this IAliveEntity aliveEntity)
+        {
+            return new PflagPacket
+            {
+                VisualType = aliveEntity.VisualType,
+                VisualId = aliveEntity.VisualId,
+                Flag = aliveEntity.Shop?.ShopId ?? 0
+            };
+        }
+
```
Comment (Elendan) [inline src/NosCore.GameObject/ComponentEntities/Extensions/AliveEntityExtension.cs]: Blank lines

File: src/NosCore.Controllers/NpcPacketController.cs [hunk 4]
```diff
@@ -135,10 +155,144 @@ public void Shopping(ShoppingPacket shoppingPacket)
...
+                return;
+            }
+
+            if (Session.Character.Group != null && Session.Character.Group?.Type != GroupType.Group)
+            {
+                Session.SendPacket(new MsgPacket
+                {
+                    Message = Language.Instance.GetMessageFromKey(LanguageKey.SHOP_NOT_ALLOWED_IN_RAID,
+                        Session.Account.Language),
+                    Type = 0
```
Comment (Elendan) [inline src/NosCore.Controllers/NpcPacketController.cs]: Is there no enum for Msg types yet ?

File: src/NosCore.Controllers/NpcPacketController.cs [hunk 5]
```diff
@@ -135,10 +155,144 @@ public void Shopping(ShoppingPacket shoppingPacket)
...
+                return;
+            }
+
+            if (!Session.Character.MapInstance.ShopAllowed)
+            {
+                Session.SendPacket(new MsgPacket
+                {
+                    Message = Language.Instance.GetMessageFromKey(LanguageKey.SHOP_NOT_ALLOWED,
+                        Session.Account.Language),
+                    Type = 0
```
Comment (Elendan) [inline src/NosCore.Controllers/NpcPacketController.cs]: same here

File: src/NosCore.Controllers/NpcPacketController.cs [hunk 6]
```diff
@@ -135,10 +155,144 @@ public void Shopping(ShoppingPacket shoppingPacket)
...
+                            //todo log
+                            return;
+                        }
+
+                        if (!inv.Item.IsTradable || inv.BoundCharacterId != null)
+                        {
+                            Session.SendPacket(Session.Character.GenerateSay(
+                                Language.Instance.GetMessageFromKey(LanguageKey.SHOP_ONLY_TRADABLE_ITEMS, Session.Account.Language),
+                                SayColorType.Yellow));
+                            Session.SendPacket(new ShopEndPacket { Type = 0 });
```
Comment (Elendan) [inline src/NosCore.Controllers/NpcPacketController.cs]: Enum here as well ?

File: src/NosCore.Controllers/NpcPacketController.cs [hunk 7]
```diff
@@ -135,10 +155,144 @@ public void Shopping(ShoppingPacket shoppingPacket)
...
+                            return;
+                        }
+
+                        Session.Character.Shop.ShopItems.TryAdd(shopSlot,
+                            new ShopItem
+                            {
+                                Amount = item.Amount,
+                                Price = item.Price,
+                                Slot = shopSlot,
+                                Type = 0,
```
Comment (Elendan) [inline src/NosCore.Controllers/NpcPacketController.cs]: Same here

File: src/NosCore.Controllers/NpcPacketController.cs [hunk 8]
```diff
@@ -135,10 +155,144 @@ public void Shopping(ShoppingPacket shoppingPacket)
...
+                                ItemInstance = inv
+                            });
+                    }
+
+                    if (Session.Character.Shop.ShopItems.Count == 0)
+                    {
+                        Session.SendPacket(Session.Character.GenerateSay(
+                            Language.Instance.GetMessageFromKey(LanguageKey.SHOP_EMPTY, Session.Account.Language),
+                            SayColorType.Yellow));
+                        Session.SendPacket(new ShopEndPacket { Type = 0 });
```
Comment (Elendan) [inline src/NosCore.Controllers/NpcPacketController.cs]: same here

File: src/NosCore.GameObject/Character.cs [hunk 9]
```diff
@@ -472,43 +486,56 @@ public void Buy(Shop shop, short slot, short amount)
 
             if (reputprice > Session.Character.Reput)
             {
-                Session.SendPacket(new SMemoPacket
+                SendPacket(new SMemoPacket
                 {
                     Type = 3,
```
Comment (Elendan) [inline src/NosCore.GameObject/Character.cs]: same here

File: src/NosCore.GameObject/Character.cs [hunk 10]
```diff
@@ -524,6 +551,52 @@ public void Buy(Shop shop, short slot, short amount)
...
+            var slot = item.Slot;
+            item.Amount -= amount;
+            if ((item?.Amount ?? 0) == 0)
+            {
+                Shop.ShopItems.TryRemove(slot, out _);
+            }
+            SendPacket(itemInstance.GeneratePocketChange(type, slotChar));
+            SendPacket(new SMemoPacket
+            {
+                Type = 1,
```
Comment (Elendan) [inline src/NosCore.GameObject/Character.cs]: same here

File: src/NosCore.GameObject/ComponentEntities/Extensions/AliveEntityExtension.cs [hunk 11]
```diff
@@ -270,20 +282,28 @@ public static void SetLevel(this INamedEntity experiencedEntity, byte level)
...
-                .OrderBy(s => s.Slot))
+            var list = aliveEntity.Shop.ShopItems.Values.Where(s => s.Type.Equals(typeshop)).ToList();
+            for (var i = 0; i < aliveEntity.Shop.Size; i++)
             {
-                shopItemList.Add(new NInvItemSubPacket
+                var item = list.Find(s=>s.Slot == i);
+                if (item == null)
+                {
+                    shopItemList.Add(null);
+                }
```
Comment (Elendan) [inline src/NosCore.GameObject/ComponentEntities/Extensions/AliveEntityExtension.cs]: this feels weird, is it normal ?

File: src/NosCore.Controllers/NpcPacketController.cs (hunk 2 above)
Comment (0Lucifer0) [reply src/NosCore.Controllers/NpcPacketController.cs]: nope, it's always 3 or the one in database. No way to explain what it is doing here

File: src/NosCore.GameObject/ComponentEntities/Extensions/AliveEntityExtension.cs (hunk 11 above)
Comment (0Lucifer0) [reply src/NosCore.GameObject/ComponentEntities/Extensions/AliveEntityExtension.cs]: yes it is. we want the item to be null to get -1

File: src/NosCore.Controllers/NpcPacketController.cs (hunk 7 above)
Comment (0Lucifer0) [reply src/NosCore.Controllers/NpcPacketController.cs]: there is no type here it's the shop id. A character can only have 1 shop so 0

File: src/NosCore.Controllers/NpcPacketController.cs (hunk 1 above)
Comment (0Lucifer0) [reply src/NosCore.Controllers/NpcPacketController.cs]: i don't think portal deserve it. it's not a visual entity

**Pull request merged**